fendermint/vm/topdown/src/sync.rs

```diff
@@ -0,0 +1,273 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+//! A constant running process that fetch or listener to parent state
+
+use crate::error::Error;
+use crate::proxy::{IPCProviderProxy, ParentQueryProxy};
+use crate::{
+    BlockHash, BlockHeight, CachedFinalityProvider, Config, IPCParentFinality,
+    ParentFinalityProvider, Toggle,
+};
+use anyhow::{anyhow, Context};
+use async_stm::{atomically, atomically_or_err};
+use ipc_sdk::cross::CrossMsg;
+use ipc_sdk::staking::StakingChangeRequest;
+use std::cmp::min;
+use std::sync::Arc;
+use std::time::Duration;
+
+/// The max number of blocks polling should query each parent view update. If the number of blocks
+/// polled equals this value, it would stop polling for this iteration and commit the result to cache.
+const MAX_PARENT_VIEW_BLOCK_GAP: BlockHeight = 100;
+/// When polling parent view, if the number of top down messages exceeds this limit,
+/// the polling will stop for this iteration and commit the result to cache.
+const TOPDOWN_MSG_LEN_THRESHOLD: usize = 500;
+
+/// Query the parent finality from the block chain state
+pub trait ParentFinalityStateQuery {
+    /// Get the latest committed finality from the state
+    fn get_latest_committed_finality(&self) -> anyhow::Result<Option<IPCParentFinality>>;
+}
+
+/// Constantly syncing with parent through polling
+pub struct PollingParentSyncer<T> {
+    config: Config,
+    parent_view_provider: Arc<Toggle<CachedFinalityProvider<IPCProviderProxy>>>,
+    agent: Arc<IPCProviderProxy>,
```

Comment:
Do we want to still call this one `agent`? Should we may be call it `ipc` instead?

Reply:
Maybe `parent_client`?

Reply:
fixed!